pkg/controller/direct/asset/savedquery_controller.go

```diff
@@ -0,0 +1,390 @@
+}
+
+func NewSavedQueryModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &savedQueryModel{config: *config}, nil
+}
+
+var _ directbase.Model = &savedQueryModel{}
+
+type savedQueryModel struct {
+	config config.ControllerConfig
+}
+
+func (m *savedQueryModel) client(ctx context.Context, projectID string) (*gcp.Client, error) {
+	var opts []option.ClientOption
+
+	config := m.config
+	// Workaround for an unusual behaviour (bug?):
+	//  the service requires that a quota project be set
+	if !config.UserProjectOverride || config.BillingProject == "" {
+		config.UserProjectOverride = true
+		config.BillingProject = projectID
+	}
+	opts, err := config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	gcpClient, err := gcp.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building asset saved query client: %w", err)
+	}
+
+	return gcpClient, err
+}
+
+func (m *savedQueryModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.AssetSavedQuery{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewSavedQueryIdentity(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	gcpClient, err := m.client(ctx, id.Parent().ProjectID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &savedQueryAdapter{
+		gcpClient: gcpClient,
+		id:        id,
+		desired:   obj,
+		reader:    reader,
+	}, nil
+}
+
+func (m *savedQueryModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type savedQueryAdapter struct {
+	gcpClient *gcp.Client
+	id        *krm.SavedQueryIdentity
+	desired   *krm.AssetSavedQuery
+	actual    *pb.SavedQuery
+	reader    client.Reader
+}
+
+var _ directbase.Adapter = &savedQueryAdapter{}
+
+func (a *savedQueryAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting asset saved query", "name", a.id)
+
+	name := a.id.String()
+	if a.actual != nil {
+		name = a.actual.Name
+	}
+	req := &pb.GetSavedQueryRequest{Name: name}
+	actual, err := a.gcpClient.GetSavedQuery(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting asset saved query %q from gcp: %w", a.id.String(), err)
+	}
+
+	a.actual = actual
+	return true, nil
+}
+
+func (a *savedQueryAdapter) normalizeReferences(ctx context.Context) error {
+	// No references to resolve in SavedQuery spec currently.
+	return nil
+}
+
+// AssetSavedQueryStatus_FromProto converts the AssetSavedQueryStatus field from its Protobuf representation.
+func AssetSavedQueryStatus_FromProto(mapCtx *direct.MapContext, in *pb.SavedQuery) *krm.AssetSavedQueryStatus {
```

Comment:
Can we move this to the mapper file? Or, can we just use the mapping function for ObservedState?

Reply:
Moved it to mapper file.